dripline/extensions/ethernet_thermo_fisher_service.py

```diff
@@ -0,0 +1,172 @@
+import re
+import socket
+import threading
+import time
+
+from dripline.core import Service, ThrowReply
+
+import logging
+logger = logging.getLogger(__name__)
+
+__all__ = []
+
+
+__all__.append('EthernetThermoFisherService')
+
+def int_to_hexstr(value):
+    return hex(value)[2:].zfill(2)
+
+def hexstr_to_bytes(value):
+    return bytes.fromhex(value)
+
+def bytes_to_hexstr(value):
+    return value.hex()
+
+def hexstr_to_int(value):
+    return int(value, 16)
+
+def bytes_to_ints(value):
+    return [byte for byte in value]
+
+class EthernetThermoFisherService(Service):
+    '''
+    A fairly specific subclass of Service for connecting to ethernet-capable thermo fisher devices.
+    In particular, devices must support a half-duplex serial communication with header information, variable length data-payload and a checksum.
+    '''
+    def __init__(self,
+                 socket_timeout=10,
+                 socket_info=('localhost', 1234),
+                 lead_char = b'\xcc',
+                 msb = b'\x00',
+                 lsb = b'\x01',
+                 **kwargs
+                 ):
+        '''
+        Args:
+            socket_timeout (int): number of seconds to wait for a reply from the device before timeout.
+            socket_info (tuple or string): either socket.socket.connect argument tuple, or string that
+                parses into one.
+        '''
+        Service.__init__(self, **kwargs)
+
+        if isinstance(socket_info, str):
+            logger.debug(f"Formatting socket_info: {socket_info}")
+            re_str = "\([\"'](\S+)[\"'], ?(\d+)\)"
+            (ip,port) = re.findall(re_str,socket_info)[0]
+            socket_info = (ip,int(port))
+
+        self.alock = threading.Lock()
+        self.socket = socket.socket()
+        self.socket_timeout = float(socket_timeout)
+        self.socket_info = socket_info
+        self.lead_char = lead_char
+        self.msb = msb
+        self.lsb = lsb
+
+    def send_to_device(self, commands, **kwargs):
```

Comment:
this is curiously close to the standard SCPI send_to_device, but doesn't exactly match
can you elaborate what is getting fixed?  and if we should consider porting some of these changes into the base class?

Reply:
I do not keep the connection open and the handling of the reconnection. Thats moved into the `_send_commands` function.